README.md
acewater
================

<!-- README.md is generated from README.Rmd. Please edit that file -->
[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active) [![license](https://img.shields.io/github/license/mashape/apistatus.svg)]()

acewater provides an interface to the real-time water monitoring data provided by the [Water Management Section of the Seattle District Corps of Engineers](http://www.nwd-wc.usace.army.mil/nws/hh/www/index.html), which covers most of Washington, Northern Idaho, and Northwestern Montana. The main function, `get_water_conditions` returns a tidy data frame (a [tibble](http://tibble.tidyverse.org)).

Installation
------------

You can install acewater from github with:

``` r
# install.packages("devtools")
devtools::install_github("briandconnelly/acewater")
```

Example
-------

### Get the current conditions at the [Grand Coulee Dam](https://en.wikipedia.org/wiki/Grand_Coulee_Dam) (GCL), part of the [Eastern Washington Rivers](http://www.nwd-wc.usace.army.mil/nws/hh/www/eastern.html) (EWA) project:

``` r
library(acewater)

(gcl <- get_water_conditions(station = "GCL", project = "EWA"))
#> # A tibble: 42 x 6
#>          Time Project Station Depth     Measure   Value
#>        <dttm>  <fctr>  <fctr> <dbl>      <fctr>   <dbl>
#>  1 2017-06-03     EWA     GCL    NA ElevForebay 1263.29
#>  2 2017-06-04     EWA     GCL    NA ElevForebay 1265.77
#>  3 2017-06-05     EWA     GCL    NA ElevForebay 1267.69
#>  4 2017-06-06     EWA     GCL    NA ElevForebay 1269.61
#>  5 2017-06-07     EWA     GCL    NA ElevForebay 1271.45
#>  6 2017-06-08     EWA     GCL    NA ElevForebay 1272.87
#>  7 2017-06-09     EWA     GCL    NA ElevForebay 1274.10
#>  8 2017-06-10     EWA     GCL    NA ElevForebay 1275.37
#>  9 2017-06-11     EWA     GCL    NA ElevForebay 1276.71
#> 10 2017-06-12     EWA     GCL    NA ElevForebay 1277.87
#> # ... with 32 more rows
```

We can easily plot the water conditions for a given station (if [ggplot2](http://ggplot2.tidyverse.org) is installed).

``` r
library(ggplot2)

autoplot(gcl)
```

![](README-Example%20Plot%20Grand%20Coulee%20Dam-1.png)

### Get the most recent conditions at the [Fremont Bridge](https://en.wikipedia.org/wiki/Fremont_Bridge_(Seattle)) (station FBLW) in the [Lake Washington Ship Canal](https://en.wikipedia.org/wiki/Lake_Washington_Ship_Canal) (project LKW):

Here, we see that this station monitors different properties:

``` r
(fblw <- get_water_conditions(station = "FBLW", project = "LKW"))
#> # A tibble: 156 x 6
#>                   Time Project Station Depth  Measure Value
#>                 <dttm>  <fctr>  <fctr> <dbl>   <fctr> <dbl>
#>  1 2017-06-15 12:00:00     LKW    FBLW    18 Salinity  0.04
#>  2 2017-06-15 13:00:00     LKW    FBLW    18 Salinity  0.04
#>  3 2017-06-15 14:00:00     LKW    FBLW    18 Salinity  0.04
#>  4 2017-06-15 15:00:00     LKW    FBLW    18 Salinity  0.04
#>  5 2017-06-15 16:00:00     LKW    FBLW    18 Salinity  0.04
#>  6 2017-06-15 17:00:00     LKW    FBLW    18 Salinity  0.04
#>  7 2017-06-15 18:00:00     LKW    FBLW    18 Salinity  0.04
#>  8 2017-06-15 19:00:00     LKW    FBLW    18 Salinity  0.04
#>  9 2017-06-15 20:00:00     LKW    FBLW    18 Salinity  0.04
#> 10 2017-06-15 21:00:00     LKW    FBLW    18 Salinity  0.04
#> # ... with 146 more rows
autoplot(fblw)
```

![](README-Example%20Fremont%20Bridge-1.png)

### Get Conditions from Multiple Stations

We can use [purrr](http://purrr.tidyverse.org) to combine data from multiple stations into a tidy data frame. Here, we'll get data from a few stations along the [Lake Washington Ship Canal](https://en.wikipedia.org/wiki/Lake_Washington_Ship_Canal).

``` r
(lkw <- purrr::map_df(c("UBLW", "FBLW", "LLLW"), get_water_conditions, project = "LKW"))
#> # A tibble: 520 x 6
#>                   Time Project Station Depth  Measure Value
#>                 <dttm>  <fctr>   <chr> <dbl>   <fctr> <dbl>
#>  1 2017-06-15 12:00:00     LKW    UBLW     8 Salinity  0.03
#>  2 2017-06-15 13:00:00     LKW    UBLW     8 Salinity  0.03
#>  3 2017-06-15 14:00:00     LKW    UBLW     8 Salinity  0.03
#>  4 2017-06-15 15:00:00     LKW    UBLW     8 Salinity  0.03
#>  5 2017-06-15 16:00:00     LKW    UBLW     8 Salinity  0.03
#>  6 2017-06-15 17:00:00     LKW    UBLW     8 Salinity  0.03
#>  7 2017-06-15 18:00:00     LKW    UBLW     8 Salinity  0.03
#>  8 2017-06-15 19:00:00     LKW    UBLW     8 Salinity  0.03
#>  9 2017-06-15 20:00:00     LKW    UBLW     8 Salinity  0.03
#> 10 2017-06-15 21:00:00     LKW    UBLW     8 Salinity  0.03
#> # ... with 510 more rows
autoplot(lkw)
```

![](README-Example%20Multiple%20Stations-1.png)

Data Subject to Revision
------------------------

Because the data retrieved by acewater are preliminary and have not been verified or approved by the agencies collecting them, inaccuracies may be present due to instrument malfunctions or physical changes at the measurement site.

Contributer Code of Conduct
---------------------------

Please note that this project is released with a [Contributor Code of Conduct](CONDUCT.md). By participating in this project you agree to abide by its terms.


NEWS.md
# acewater NEXT

* Added `autoplot.ace_conditions` and `autoplot.ace_station` to create plots using ggplot2


# acewater 0.4.0 (2017-06-09)

* Added stations for all Seattle District projects: Chehalis River Basin, Eastern Washington Rivers, Flathead and Clark Fork Rivers, Green River Basin, Kootenai River Basin, Pend Oreille River Basin, Puyallup River Basin, and Skagit River Basin
* Updated `ace_station` structure
    * Added `data_units` field to `ace_station` for measurement units
    * Added `project` field to `ace_station` to store station's project
    * Added `info_url` field for adding links to external information
    * Added `latitude` and `longitude` fields for mapping purposes [#5]
* Added `ace_project` structure for storing projects (each has multiple stations)
* Added `find_project` function to find the project (code) associated with a given monitoring station
* `get_station` now takes a project code as argument
* Updated `get_water_conditions`
    * `get_water_conditions.character` now takes a project code as argument
    * Project/basin now included in result tibble
    * Additional attributres like project/station name and retrieval time added [#6]
    * Added `ace_conditions` class to allow for future options like `plot.ace_conditions`
* Removed `station_name` function
* Internal data updates
    * Added values for `data_units` and `project` to existing stations
    * Added `ace_projects` data set, containing information about different basins/projects
    * Removed `ace_stations` (after renaming it from stations)
    * File moved to `ace_projects`
* README updates
    * Added Grand Coulee Dam example
    * Updated examples in README to include project codes




# acewater 0.3.0 (2017-06-07)

* Now works with all stations in the Lake Washington Ship Canal
* `get_water_conditions` can now work with more types of measurements (just using the data_columns names)
* `get_water_conditions` now returns Station code as a factor
* Added more stations to stations data set: BTHS, FISH2, KNMR, RNTW
* Added `data_skiprows` attribute to `ace_station` to specify number of rows to skip at the beginning of the file


# acewater 0.2.1 (2017-06-07)

* Added `ace_station` constructor
    * stations data is now a list of ace_station objects
    * `get_station` now returns object from station data, no longer constructing an object itself
    * `get_water_conditions` now works with `ace_station` objects
* `get_water_conditions` now a generic function
    * Supports for specifying station as character vector (e.g. "FBLW") or `ace_station` object


# acewater 0.2.0 (2017-06-07)

* get_water_conditions now takes station code instead of station object
* Station data now includes information about columns in result, which will make it easy to accomodate more station data formats
* Station data can now include a url in case a station's url doesn't follow the usual scheme
* Moved station data for stations.R to data-raw/data_stations.R
* Station data are know stored as package data (stations)
* `get_station` now returns an `ace_station` object
* Added `station_name` helper function
* Added basic unit testing framework


# acewater 0.1.0

* Initial version
* Added a `NEWS.md` file to track changes to the package.


data-raw/data_stations.R
# ace_project contains a list of project names and their codes
ace_projects <- list(
    "chh" = "Chehalis River Basin",
    "ewa" = "Eastern Washington Rivers",
    "flt" = "Flathead and Clark Fork Rivers",
    "grn" = "Green River Basin",
    "kot" = "Kootenai River Basin",
    "lkw" = "Lake Washington",
    "pnd" = "Pend Oreille River Basin",
    "puy" = "Puyallup River Basin",
    "skg" = "Skagit River Basin"
)


# Stations contains information about known monitoring stations
ace_stations <- list(
    "BBLW" = ace_station(
        name = "Ballard Bridge",
        code = "BBLW",
        data_columns = c("Time",
                         "SalinityA",
                         "SalinityB",
                         "SalinityC",
                         "TemperatureA",
                         "TemperatureB",
                         "TemperatureC"),
        depths = c("A" = 11, "B" = 21, "C" = 32),
        data_skiprows = 5
    ),
    "FBLW" = ace_station(
        name = "Fremont Bridge",
        code = "FBLW",
        data_columns = c("Time",
                         "SalinityA",
                         "SalinityB",
                         "SalinityC",
                         "TemperatureA",
                         "TemperatureB",
                         "TemperatureC"),
        depths = c("A" = 18, "B" = 31, "C" = 40),
        data_skiprows = 5
    ),
    "UBLW" = ace_station(
        name = "University Bridge",
        code = "UBLW",
        data_columns = c("Time",
                         "SalinityA",
                         "SalinityB",
                         "SalinityC",
                         "TemperatureA",
                         "TemperatureB",
                         "TemperatureC"),
        depths = c("A" = 8, "B" = 21, "C" = 35),
        data_skiprows = 5
    ),
    "GWLW" = ace_station(
        name = "Gas Works Park",
        code = "GWLW",
        data_columns = c("Time",
                         "SalinityA",
                         "SalinityB",
                         "SalinityC",
                         "SalinityD",
                         "TemperatureA",
                         "TemperatureB",
                         "TemperatureC",
                         "TemperatureD"),
        depths = c("A" = 5, "B" = 14, "C" = 29, "D" = 44),
        data_skiprows = 5
    ),
    "LLLW" = ace_station(
        name = "Large Lock",
        code = "LLLW",
        data_columns = c("Time",
                         "SalinityA",
                         "SalinityB",
                         "SalinityC",
                         "SalinityD",
                         "TemperatureA",
                         "TemperatureB",
                         "TemperatureC",
                         "TemperatureD"),
        depths = c("A" = 18, "B" = 28, "C" = 36, "D" = 43),
        data_skiprows = 5
    ),
    "FISH2" = ace_station(
        name = "Fish Ladder",
        code = "FISH2",
        data_columns = c("Time",
                         "Conductivity",
                         "Salinity",
                         "Temperature",
                         "FishUp",
                         "FishDn"),
        depths = c(),
        data_skiprows = 6
    ),
    "BTHS" = ace_station(
        name = "Locks Boathouse",
        code = "BTHS",
        data_columns = c("Time",
                         "Elevation"),
        depths = c(),
        data_skiprows = 6
    ),
    "KNMR" = ace_station(
        name = "Kenmore Gage",
        code = "KNMR",
        data_columns = c("Time",
                         "Elevation"),
        depths = c(),
        data_skiprows = 5
    ),
    "RNTW" = ace_station(
        name = "Cedar River at Renton",
        code = "RNTW",
        data_columns = c("Time",
                         "Stage",
                         "Flow"),
        depths = c(),
        data_skiprows = 5
    )
)

devtools::use_data(ace_projects, ace_stations, internal = TRUE, overwrite = TRUE)
rm(ace_projects, ace_stations)
